### DISCUSSION TITLE
[Question] Bluetooth socket recv

### DISCUSSION THREAD

**Discussion by alkebuware** [halildurmus/win32#318] [Q&A]
Hi Tim,

Thanks for all the effort you've put into this project. It's some great work. I do have a question about the bluetooth  and socket level programming using this library, and was hoping you could give me some pointers. 

Using the functions and classes in bthprops I'm able to succesfully ascertain all of the connected devices to a Windows machine. However, I'm having the hardest time figuring out how to connect to a bluetooth device.

I'm able to successfully create a socket using the below block of code using dart:ffi.
```dart
final socketLibrary = DynamicLibrary.open('ws2_32.dll');
  final socket = socketLibrary.lookupFunction<
      Int32 Function(Int32 af, Int32 type, Int32 protocol),
      int Function(int af, int type, int protool)>('socket');
  final s = socket(/*AF_BTH*/ 32, /*SOCK_STREAM*/ 1, /*BTHPROTO_RFCOMM*/ 3);
```

However, when I try to `connect` to the socket I keep getting an error code: -1. Below is my current set up for attempting to connect to the socket. I've hardcoded the address, and serviceClassId to the GUID returned by `BluetoothEnumerateInstalledServices`. 
```dart
class SOCKADDR_BTH extends Struct {
  @Uint16()
  external int addressFamily;
  external BLUETOOTH_ADDRESS btAddr;
  external GUID serviceClassId;
  @Uint32()
  external int port;
}

 final address = calloc<SOCKADDR_BTH>()
    ..ref.port = 0
    ..ref.addressFamily = 32 /*AF_BTH*/
    ..ref.btAddr.ullLong = 71751828756671
    ..ref.serviceClassId.setGUID("{0000111E-0000-1000-8000-00805F9B34FB}");
  final result = connect(s, address, sizeOf<SOCKADDR_BTH>());
```

The endgame is to be able to read or `recv` from the socket incoming data from the bluetooth device. Any help is greatly appreciated.

**Comment by timsneath** [comment]
Hmm.... I don't _see_ anything obviously wrong, but I've done little sockets programming before. Is there a way I can repro this somehow and see the error you're getting? 

I have now [added sockets API support to win32](https://github.com/timsneath/win32/pull/260) on the builds in the `main` channel. Because of the high risk of name clash with other libraries, you'll need to import it separately, with something like the following:

```dart
import 'package:win32/win32.dart';
import 'package:win32/winsock2.dart' as winsock2;
```

Then you can call `winsock2.connect()` or `winsock2.send()` without needing to manually declare your the APIs or structs.

**Comment by AErmek** [comment]
> Hi Tim,
> 
> Thanks for all the effort you've put into this project. It's some great work. I do have a question about the bluetooth and socket level programming using this library, and was hoping you could give me some pointers.
> 
> Using the functions and classes in bthprops I'm able to succesfully ascertain all of the connected devices to a Windows machine. However, I'm having the hardest time figuring out how to connect to a bluetooth device.
> 
> I'm able to successfully create a socket using the below block of code using dart:ffi.
> 
> ```dart
> final socketLibrary = DynamicLibrary.open('ws2_32.dll');
>   final socket = socketLibrary.lookupFunction<
>       Int32 Function(Int32 af, Int32 type, Int32 protocol),
>       int Function(int af, int type, int protool)>('socket');
>   final s = socket(/*AF_BTH*/ 32, /*SOCK_STREAM*/ 1, /*BTHPROTO_RFCOMM*/ 3);
> ```
> 
> However, when I try to `connect` to the socket I keep getting an error code: -1. Below is my current set up for attempting to connect to the socket. I've hardcoded the address, and serviceClassId to the GUID returned by `BluetoothEnumerateInstalledServices`.
> 
> ```dart
> class SOCKADDR_BTH extends Struct {
>   @Uint16()
>   external int addressFamily;
>   external BLUETOOTH_ADDRESS btAddr;
>   external GUID serviceClassId;
>   @Uint32()
>   external int port;
> }
> 
>  final address = calloc<SOCKADDR_BTH>()
>     ..ref.port = 0
>     ..ref.addressFamily = 32 /*AF_BTH*/
>     ..ref.btAddr.ullLong = 71751828756671
>     ..ref.serviceClassId.setGUID("{0000111E-0000-1000-8000-00805F9B34FB}");
>   final result = connect(s, address, sizeOf<SOCKADDR_BTH>());
> ```
> 
> The endgame is to be able to read or `recv` from the socket incoming data from the bluetooth device. Any help is greatly appreciated.

Hi, 
did you find any solution? i also get `-1` when `connect` and `WSAGetLastError ` returned **10049** (WSAEADDRNOTAVAIL) error
I've tried many combinations of `btAddr`,`serviceClassId`,`port` but always `-1`.

@timsneath thank you for package, 
but i'm little confused, `winsock2.connect` methods second parameter is `Pointer<SOCKADDR>`, but i should pass there `Pointer<SOCKADDR_BTH>`, is this normal  to use `cast` method from `ffi`?
```dart
final address = calloc<SOCKADDR_BTH>()
    ..ref.addressFamily = winsock2.AF_BTH
    ..ref.btAddr = deviceAddress
    ..ref.serviceClassId.setGUID(serviceId)
    ..ref.port = BT_PORT_ANY;
    
winsock2.connect(
      bluetoothSocket,
      address.cast<winsock2.SOCKADDR>(),
      sizeOf<SOCKADDR_BTH>(),
    );
```
